### PR TITLE
Change Apex tests from Xunit to MSTest

### PR DESCRIPTION
**Pull request by heng-liu**
## Bug

Fixes: https://github.com/nuget/client.engineering/issues/852

Regression? Last working version:

## Description
**Background**
Since Xunit is not able to timeout individual test ([comment](https://github.com/nuget/client.engineering/issues/852#issuecomment-875681103)), in order to time out individual Apex tests, change Apex tests from Xunit to MSTest.

1.Change Apex test from Xunit to MSTest:
   * Change Xunit attributes to MSTest attributes: `Theory `-> `TestMethod`, `InlineData `-> `DataRow`, `Skip `-> `Ignore`, etc.
   * Add `Timeout `attribute for each Apex tests
   * Change Xunit.Assert to MSTest assertion 
   * Adapt setup of Xunit (fixture) to MSTest setup (ClassInitialize). (MSTest has [limitation ](https://github.com/microsoft/testfx/issues/143) as ClassInitialize is not inheritable. So change the previous fixture for base test class to a static property.)
   * Change Xunit log into Trace.WriteLine. (a test [build ](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=5326490&view=ms.vss-test-web.build-test-results-tab&runId=27181036&resultId=100007&paneView=attachments) shows that when the test failed, the log will be shown)
 2. Engineering change 
 * Add package for MSTest: MSTest.TestAdapter and MSTest.TestFramework
 * Change build/template.runsettingsproj to add console logger, so that trace will be shown in the log.
 * Clean up test/NuGet.Tests.Apex/NuGet.Tests.Apex/xunit.runner.json
 * Decide not to clean up xunit packages from Apex test projects, as Apex test projects reference Test.Utility project, which has xunit package reference. We will have to get rid of all references to xunit from Test.Utility project first. I don't think it's worth doing compared to the effort (The only impact is we still see several xunit message in test logs).

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [x] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: build/packages.targets [hunk 1]
```diff
@@ -113,6 +113,8 @@
         <PackageReference Update="xunit" Version="$(XunitVersion)" />
         <PackageReference Update="xunit.runner.visualstudio" Version="$(XunitVersion)" />
         <PackageReference Update="Xunit.StaFact" Version="0.2.9" />
+        <PackageReference Update="MSTest.TestAdapter" Version="2.1.1" />
+        <PackageReference Update="MSTest.TestFramework" Version="2.1.1" />
```
Comment (erdembayar) [inline build/packages.targets:117]: Why this version 2.2.1 is chosen? It looks latest version is [2.2.7](https://www.nuget.org/packages/MSTest.TestFramework/2.2.7).

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NetCoreProjectTestCase.cs [hunk 2]
```diff
@@ -1,41 +1,45 @@
...
 
 namespace NuGet.Tests.Apex
 {
-    public class NetCoreProjectTestCase : SharedVisualStudioHostTestClass, IClassFixture<VisualStudioHostFixtureFactory>
+    [TestClass]
+    public class NetCoreProjectTestCase : SharedVisualStudioHostTestClass
     {
-        public NetCoreProjectTestCase(VisualStudioHostFixtureFactory visualStudioHostFixtureFactory, ITestOutputHelper output)
-            : base(visualStudioHostFixtureFactory, output)
+        private const int Timeout = 5 * 60 * 1000; // 5 minutes
```
Comment (erdembayar) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NetCoreProjectTestCase.cs:10]: Maybe move this `SharedVisualStudioHostTestClass.cs` it looks repeated.

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NetCoreProjectTestCase.cs [hunk 3]
```diff
@@ -1,41 +1,45 @@
...
+        public NetCoreProjectTestCase()
+            : base()
         {
         }
 
         // basic create for .net core template
-        [NuGetWpfTheory]
-        [MemberData(nameof(GetNetCoreTemplates))]
-        public void CreateNetCoreProject_RestoresNewProject(ProjectTemplate projectTemplate)
+        [TestMethod]
```
Comment (erdembayar) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NetCoreProjectTestCase.cs:18]: Do we need to change `NuGetWpfTheory` to MSTest attribute or created new one?
https://github.com/NuGet/NuGet.Client/blob/892375745964db627494305fd605496454b5ad0a/test/TestExtensions/NuGet.StaFact/NuGetWpfTheoryAttribute.cs#L10

**Review by erdembayar — COMMENTED** [review]
Start with few comments. In general looks good.

**Comment by kartheekp-ms** [comment]
If I am reading https://github.com/xunit/xunit/issues/217#issuecomment-625402054 from the `xUnit` maintainer correctly, the framework supports timeout parameter for `async` tests. May we should change one `async` test locally and check if timeout works.

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/Apex/ApexTestContext.cs [hunk 4]
```diff
@@ -22,9 +21,9 @@ internal class ApexTestContext : IDisposable
 
         public NuGetApexTestService NuGetApexTestService { get; }
 
-        public ApexTestContext(VisualStudioHost visualStudio, ProjectTemplate projectTemplate, ILogger logger, bool noAutoRestore = false, bool addNetStandardFeeds = false, SimpleTestPathContext simpleTestPathContext = null)
+        public ApexTestContext(VisualStudioHost visualStudio, ProjectTemplate projectTemplate, bool noAutoRestore = false, bool addNetStandardFeeds = false, SimpleTestPathContext simpleTestPathContext = null)
         {
-            logger.LogInformation("Creating test context");
+            Trace.WriteLine("Creating test context");
```
Comment (erdembayar) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/Apex/ApexTestContext.cs:26]: How we're monitoring this message? If I remember correctly then it writes to VS "Output window". How do we check this?

**Comment by kartheekp-ms** [comment]
> If I am reading [xunit/xunit#217 (comment)](https://github.com/xunit/xunit/issues/217#issuecomment-625402054) from the `xUnit` maintainer correctly, the framework supports timeout parameter for `async` tests. May we should change one `async` test locally and check if timeout works.

I just running an apex test locally with a timeout and disabling parallel execution of tests. Unfortunately, tests didn't stop after timeout.

**Comment by heng-liu** [comment]
> If I am reading [xunit/xunit#217 (comment)](https://github.com/xunit/xunit/issues/217#issuecomment-625402054) from the `xUnit` maintainer correctly, the framework supports timeout parameter for `async` tests. May we should change one `async` test locally and check if timeout works.

Thanks @kartheekp-ms ! I checked the issue, too. But there are two Apex test classes [IVsServicesTestCase](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/IVsServicesTestCase.cs) and [NetCoreProjectTestCase](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NetCoreProjectTestCase.cs), having not async test cases, which will still not be able to timed out if using Xunit.

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/Apex/ApexTestContext.cs (hunk 4 above)
Comment (heng-liu) [reply test/NuGet.Tests.Apex/NuGet.Tests.Apex/Apex/ApexTestContext.cs:26]: Sorry for the late response..
I think `logger.LogInformation` is not writing to VS "Output window", but print log to the testing console.
If checking the following code:
1.ApexTestContext accept ILogger logger in the constructor at [here](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Tests.Apex/NuGet.Tests.Apex/Apex/ApexTestContext.cs#L25)
2.One example of new a ApexTestContext object is [this](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NetCoreProjectTestCase.cs#L105)
3.And the XunitLogger is from [here](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Tests.Apex/NuGet.Tests.Apex/Apex/SharedVisualStudioHostTestClass.cs#L39)
4.If checking the constructor of XunitLogger, it's [here](https://github.com/NuGet/NuGet.Client/blob/dev/test/TestUtilities/Test.Utility/XunitLogger.cs#L16-L20)
We need the `ITestOutputHelper ` as in Xunit, there seems no other way to print log in the testing console.
Pls check https://xunit.net/docs/capturing-output

For MSTest, we could use Trace.WriteLine to log, the output is written to an instance of DefaultTraceListener.
And in the testing, we could see the trace is shown when test failed in [this build](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=5326490&view=ms.vss-test-web.build-test-results-tab&runId=27181036&resultId=100007&paneView=attachments)

**Comment by heng-liu** [comment]
Found a related tickets I raised at that time: https://dev.azure.com/mseng/AzureDevOps/_workitems/edit/1885534

**Pull request closed without merging**